pkg/ingester/checkpoint.go

```diff
@@ -59,3 +76,420 @@ func fromWireChunks(conf *Config, wireChunks []Chunk) ([]*chunkDesc, error) {
 	}
 	return descs, nil
 }
+
+// nolint:interfacer
+func decodeCheckpointRecord(rec []byte, s *Series) error {
+	//TODO(owen-d): reduce allocs
+	// The proto unmarshaling code will retain references to the underlying []byte it's passed
+	// in order to reduce allocs. This is harmful to us because when reading from a WAL, the []byte
+	// is only guaranteed to be valid between calls to Next().
+	// Therefore, we copy it to avoid this problem.
+	cpy := make([]byte, len(rec))
+	copy(cpy, rec)
+
+	switch RecordType(cpy[0]) {
+	case CheckpointRecord:
+		return proto.Unmarshal(cpy[1:], s)
+	default:
+		return errors.Errorf("unexpected record type: %d", rec[0])
+	}
+}
+
+func encodeWithTypeHeader(m proto.Message, typ RecordType, b []byte) ([]byte, error) {
+	buf, err := proto.Marshal(m)
+	if err != nil {
+		return b, err
+	}
+
+	b = append(b[:0], byte(typ))
+	b = append(b, buf...)
+	return b, nil
+}
+
+type SeriesWithErr struct {
+	Err    error
+	Series *Series
+}
+
+type SeriesIter interface {
+	Num() int
+	Iter() <-chan *SeriesWithErr
+	Stop()
+}
+
+type ingesterSeriesIter struct {
+	ing *Ingester
+
+	done chan struct{}
+}
+
+func newIngesterSeriesIter(ing *Ingester) *ingesterSeriesIter {
+	return &ingesterSeriesIter{
+		ing:  ing,
+		done: make(chan struct{}),
+	}
+}
+
+func (i *ingesterSeriesIter) Num() (ct int) {
+	for _, inst := range i.ing.getInstances() {
+		ct += inst.numStreams()
+	}
+	return ct
+}
+
+func (i *ingesterSeriesIter) Stop() {
+	close(i.done)
+}
+
+func (i *ingesterSeriesIter) Iter() <-chan *SeriesWithErr {
+	ch := make(chan *SeriesWithErr)
+	go func() {
+		for _, inst := range i.ing.getInstances() {
+			inst.streamsMtx.RLock()
+			// Need to buffer streams internally so the read lock isn't held trying to write to a blocked channel.
+			streams := make([]*stream, 0, len(inst.streams))
+			inst.streamsMtx.RUnlock()
+			_ = inst.forAllStreams(func(stream *stream) error {
+				streams = append(streams, stream)
+				return nil
+			})
+
+			for _, stream := range streams {
+				// TODO(owen-d): use a pool
+				chunks, err := toWireChunks(stream.chunks, nil)
+				var s *Series
+				if err == nil {
+					s = &Series{
+						UserID:      inst.instanceID,
+						Fingerprint: uint64(stream.fp),
+						Labels:      client.FromLabelsToLabelAdapters(stream.labels),
+						Chunks:      chunks,
+					}
+				}
+				select {
+				case ch <- &SeriesWithErr{
+					Err:    err,
+					Series: s,
+				}:
+				case <-i.done:
+					return
+				}
+			}
+		}
+		close(ch)
+	}()
+	return ch
+}
+
+type CheckpointWriter interface {
+	// Advances current checkpoint, can also signal a no-op.
+	Advance() (noop bool, err error)
+	Write(*Series) error
+	// Closes current checkpoint.
+	Close(abort bool) error
+}
+
+type WALCheckpointWriter struct {
+	metrics    *ingesterMetrics
```

Comment:
I know this is not easy to test because it's tied to the filesystem. But I think we should still have some tests for following 300 lines of code.